eclipse-platform/eclipse.platform.swt
# Rescaling of widgets on DPI zoom changes for basic controls (win32)

Hello everyone,

**DISCLAIMER:** We are currently testing it in more complex application to identify blockers, that could result in changes in the PR. **BUT**, except from the ongoing testing the feature set in this PR includes everything we target for this first PR.
We used the branch nmodi/DynamicDPI_4909 as source of ideas and a starting point for our approach.

We are working on improving the HiDPI behaviour in SWT in the implementation for Windows for quite some time now together with @HeikoKlare. We try to split the adaptions up into smaller contributions to make the PRs easier to review and to discuss. This is the smallest set we managed to bundle having a usuable state of SWT for a subset of widgets. 
Due to this, the PR is not yet suitable to provide a proper scaling for the Eclipse IDE as a whole - some controls will not be scaled correctly or you will experience side effects.

**How to test**
As said above, the PR is just the first step, so testing it with the Eclipse IDE is not really a good idea (except to rule out side effect when **not** activating the feature. Best way to play around with the PR is to use the _ControlExample_ with the VM arguments: _-Dswt.autoScale.updateOnRuntime=true -Dswt.autoScale=quarter_. Still expect Glitches in the UI like cut off texts after DPI changes.

In the following I try to give a structured insight in our approach:

**Limitations:** (that will be covered separately in other PRs)
- With this PR there is still a limitation for one zoom level per application (`DPIUtil.deviceZoom`, see #131)
- Widgets in the common-package (e.g. CTabFolder) are not yet contained in this PR (except Item)
- There are multiple static attributes or attributes that e.g. are initialized on startup with the primary monitor zoom level and are currently not updated
 
**Opt-in feature configuration:**
The new behaviour is implemented as opt-in feature activated with the _swt.autoScale.updateOnRuntime_ flag. The logic how the DPI level is calculated is unchanged. In Windows using it with _-Dswt.autoScale=quarter_ could make sense, because the recent Windows versions usually propose scaling in 25% steps.

**DPI zoom update propagation:**
The whole propagation is integrated into the event handling triggered by `Control::WM_DPICHANGED`. The event triggers a listener in `Shell::handleZoomEvent` that propagates it into a custom compoanont with a lifecycle managed by `DPIZoomChangeRegistry`. We evaluated multiple ideas for that, like
- Adding a method directly inside _Widget_, that takes care of neccessary updates. This method would be needed to be overriden by sublassed to add additional update logic. Two major drawback for that were:
	- Exposing an additional method into the API of Widget (it must be public to be accessible from all neccessary loocations)
	- Unclear how to integrate this (Windows specific) behaviour into the common widgets, like CTabFolder
- Propagating the SWT event for a zoom change through each widget in the hierarchy and bind the update behaviour via listener (similar to `Shell::handleZoomEvent` for the "main" event). Two major drawback for that were:
	- How to register the listener? Doing it in the constructor leads to problems to distinguish if a listener must be added or not. If a Composite would be instantiated directly, you would need to register it, but not, when a subclass of it was instantiated. We didn't came a with a reliable solution to cover all issues regarding this approach.
	- How to structure the callbacks? One implementation per class that is individually called or having the listeners match the class hierarchy of the widget to delegate the event directly in the listener. Both approaches had not solvable drawback either caused by similar issues like in point 1 or by having unsorted handling of events.

In the end we decided to implement an independent workflow to have full control over the order how the updates are handled. The execution order is starting with the most general class in the class hierarchy, e.g. if a Composite is updated, the handlers are executed like (Widget -> Control -> Scrollable -> Composite). This way each class can provide a handler that only takes care for the attributes/method it adds, but the order is reliable and consistent.

**Fonts:**
With fonts there are two main challenges: How to scale fonts and how to make sure fonts are used in the correct scale?
_Scaling fonts_
The main issue arises from the conversion of pixels to points of a font. The conversion is dependent on the primary monitor zoom (see `Device::computePoints(LOGFONT logFont, long hFont)`). Therefor we need additional informartion to adapt this conversion to the correct target zoom. We achieve this by adding a package protected attribute `zoomLevel` to fonts, that is used to correct this conversion. According to our testing we see this approach works reliable.
_Identify the correct scaled variant of a font_
As a font (_instance_) can and is usually reused for many controls, we must assume these controls will be rendered on monitors with different zoom levels. Each font (_instance_) stores the handle that links to the font resource in the OS. In case a control is affected by a zoom level change, this change must be applied to the assigned font as well -> _Solution:_ update the font to a scaled variant and pass it to the control. This would be sufficient, if SWT would have the full control of font creation. But fonts are almost always created (and commonly cached) outside of SWT and we cannot assume, that a font is scaled to the correct size when passed to a control. If we would not consider this as well, e.g. a scaled font can be replaced by a non-scaled variant at any time. To cover both scenarios we adapt fonts in two places:
- Zoom level change of the Control will update the font (if != null) as well
- Update the font (usually via _setFont_) must check for the correct zoom level and replace it by a proper one, if necessary.

Both scenarios make use of (1) `Font:scaleFor(zoomLevel)` to derive scaled font variants from each font and (2) caching them in `ScalingFontRegistry` to be able to reuse scaled variants of a font that is used for multiple controls.

_Note:_ In default or if _swt.autoScale.updateOnRuntime_ flag is set to false DefaultSWTFontRegistry is used instead of `ScalingFontRegistry`

**Images:**
There was already automatic scaling in Image. This was adapted to cover additional issues we faced when testing. To reuse an image for multiple zoom levels at the same time, there will be additional adaptions necessary, according to our testing so far.

Any feedback is highly appreciated. We hope we can provide a proper starting point for improving the HiDPI behaviour  in Windows and are planning to continue with this topic in the following week.
